### ISSUE TITLE
char-rnn-classification parameters update method confusion

### ISSUE THREAD

**Issue by sevenseablue** [spro/practical-pytorch#9]
# the origin parameters update code
```python
learning_rate = 0.005
def train(category_tensor, line_tensor):
    ...
    for p in rnn.parameters():
        p.data.add_(-learning_rate, p.grad.data)
    ...
```

# I tried optimizer.step()
why do not use the optimizer.step()?
I tried optimizer.step()
```python
learning_rate = 0.005
optimizer = torch.optim.SGD(rnn.parameters(), lr=learning_rate)

def train(category_tensor, line_tensor):
    ...
    optimizer.step()
    ...
```

# result
i get a different result, loss is big , and loss flow up and down,  do not diminish.

is my code of optimizer wrong?

**Comment by spro** [comment]
I got the same results, and realized that I have the model instantiated twice - once directly after the RNN class, and once before the main `for epoch in ...` loop (below the train function). Try moving the optimizer to below the second one, so the params aren't reset.

**Comment by spro** [comment]
Should be good with 67ac021c6525347c9d69f5a6e323d935ddaf53d1
